clippy_lints/src/manual_clamp.rs

```diff
@@ -0,0 +1,785 @@
+        };
+        if let ExprKind::Call(
+            Expr { kind: ExprKind::Path(path), hir_id, .. }, &[ref arg_first, ref arg_second]
+        ) = &expr.kind
+        && cx.qpath_res(path, *hir_id) == first_res {
+            if let ExprKind::Call(
+                Expr { kind: ExprKind::Path(path_first), hir_id, .. }, &[ref arg_third, ref arg_fourth]
+            ) = &simplify_kind(arg_first).kind
+            && cx.qpath_res(path_first, *hir_id) == second_res {
+                *mut_one = Some(arg_second);
+                process_consts(arg_third, arg_fourth);
+            } else if let ExprKind::Call(
+                Expr { kind: ExprKind::Path(path_second), hir_id, .. }, &[ref arg_third, ref arg_fourth]
+            ) = &simplify_kind(arg_second).kind
+            && cx.qpath_res(path_second, *hir_id) == second_res {
+                *mut_one = Some(arg_first);
+                process_consts(arg_third, arg_fourth);
+            }
+        }
+        let input = input?;
+        let min = min?;
+        let max = max?;
+        Some(ClampSuggestion {
+            params: InputMinMax {
+                input,
+                min,
+                max,
+                is_float,
+            },
+            span: expr.span,
+            make_assignment: None,
+        })
+    }
+    // Lots of expensive queries coming up which we don't want to repeat for every expr, so short
+    // circuit this a bit.
+    let is_dual_call = |e| {
+        matches!(
+            &simplify_kind(e).kind,
+            ExprKind::Call(
+                Expr {
+                    kind: ExprKind::Path(_),
+                    ..
+                },
+                &[_, _]
+            )
+        )
+    };
+    let type_clampability = if let ExprKind::Call(
+        Expr {
+            kind: ExprKind::Path(_),
+            ..
+        },
+        &[ref arg_first, ref arg_second],
+    ) = &expr.kind
+    && (is_dual_call(arg_first) ^ is_dual_call(arg_second)) {
+        TypeClampability::is_clampable(cx, cx.typeck_results().expr_ty(arg_first))
+    } else {
+        None
+    }?;
+    let is_float = type_clampability.is_float();
+    let for_each_combo_of = |min, max| {
+        for_methods(cx, expr, true, max, min, is_float).or_else(|| for_methods(cx, expr, false, min, max, is_float))
+    };
+    if is_float {
+        let (f32_min_res, f32_max_res) = for_min_max(|m| def_path_res(cx, &["f32", m]));
+        let (f64_min_res, f64_max_res) = for_min_max(|m| def_path_res(cx, &["f64", m]));
+        for_each_combo_of(f32_min_res, f32_max_res).or_else(|| for_each_combo_of(f64_min_res, f64_max_res))
+    } else {
+        let (core_min_res, core_max_res) = for_min_max(|m| def_path_res(cx, &["core", "cmp", m]));
+        let (core_ord_min_res, core_ord_max_res) = for_min_max(|m| def_path_res(cx, &["core", "cmp", "Ord", m]));
+        for_each_combo_of(core_min_res, core_max_res).or_else(|| for_each_combo_of(core_ord_min_res, core_ord_max_res))
+    }
+}
```

Comment:
We can do something similar here, though not quite as simple as the `is_max_min_pattern` one

It would still need an inner function to test `f(g(a, b), c)` / `f(c, g(a, b))`

Something like:

```rust
/// Returns the last path segment of `func` if it's a path to [std::cmp::min], [std::cmp::max], or
/// an item of [Ord], [f32] or [f64]
fn segment<'tcx>(cx: &LateContext<'_>, func: &Expr<'tcx>) -> Option<&'tcx PathSegment<'tcx>> {
    match func.kind {
        ExprKind::Path(QPath::Resolved(None, path)) => {
            let id = path.res.opt_def_id()?;

            if matches!(cx.tcx.get_diagnostic_name(id), Some(sym::cmp_min | sym::cmp_max))
                || is_diag_trait_item(cx, id, sym::Ord)
            {
                return path.segments.last();
            }
        },
        ExprKind::Path(QPath::TypeRelative(ty, seg)) => {
            if matches!(path_res(cx, ty), Res::PrimTy(PrimTy::Float(_))) {
                return Some(seg);
            }
        },
        _ => {},
    }

    None
}

fn check<'tcx>(....) -> Option<ClampSuggestion<'tcx>> {
    if let ExprKind::call(inner_fn, .....) = outer_fn.kind
        && let Some(inner_seg) = segment(inner_fn)
        && let Some(outer_seg) = segment(outer_fn)
    // ...
    
    match (inner_seg.ident.as_str(), outer_seg.ident.as_str()) {
        ("min", "max") => {}
        ("max", "min") => {}
        _ => return None,
    }
}

if let ExprKind::Call(outer_fn, [first, second]) = &expr.kind {
    check(outer_fn, first, second).or_else(|| check(outer_fn, second, first))
} else {
    None
}
```

Reply:
This fails the test cases as written because the imported name for those was `cmp_max` and `cmp_min` in the tests. Should I change the tests? I wrote them that way because otherwise `max` and `min` were overridden by the local variables called `max` and `min`.

Reply:
However, I am able to get it to pass the tests with a little refactoring. I've just uploaded my version, let me know what you think. [`1788312`](https://github.com/rust-lang/rust-clippy/pull/9484/commits/1788312e01d0d7f882d14ed8faf110ec24cb788a)

Reply:
Ahh, didn't think of that. Yeah that change makes sense 👍